docs/docs/features/endpoints.md

```diff
@@ -5,6 +5,7 @@ title: Endpoints
 
 OAuth2 Proxy responds directly to the following endpoints. All other endpoints will be proxied upstream when authenticated. The `/oauth2` prefix can be changed with the `--proxy-prefix` config variable.
 
+- / - the proxy endpoint provides authentication and returns the appropriate 40x error if not authenticated or authorized then passes the request upstream.
```

Comment:
I don't think this is correct. You mean most likely mean `/oauth2/auth` which is the "default" endpoint for oauth2-proxy and used by your traefik / load balancer for authenticating requests.

Reply:
Would you be so kind to share your traefik and oauth2-proxy config for your local testing? So I can retest it as well?

Reply:
`oauth2/auth` doesn't work with traefik for a couple of reasons I'm using the [static upstreams configuration[https://oauth2-proxy.github.io/oauth2-proxy/configuration/integration/#forwardauth-with-static-upstreams-configuration) which uses `/` and not `/oauth2/auth`  and has uses `--upstream=static://202` to return a static 202 message.

so my traefik middleware looks something like this
```
    oauth-auth-redirect-household:
      forwardAuth:
        trustForwardHeader: true
        address: "http://oauth2-proxy?allowed_groups=household"
        authResponseHeaders:
          - X-Auth-Request-Access-Token
          - x-auth-request-preferred-username
```
 with an oauth2_proxy.cfg like this
 ```
 pass_user_headers = true
set_xauthrequest = true
pass_host_header = true
pass_authorization_header = true
set_authorization_header = true
pass_access_token = true
reverse_proxy = true
http_address = "0.0.0.0:80"

provider = "oidc"
user_id_claim = "sub" 
provider_display_name = "provider"
redirect_url = "https://login.<domain>/oauth2/callback"
oidc_issuer_url = "https://auth.<domain>"
client_id = "<clientid>"
client_secret = "<secret>"
code_challenge_method = "S256"
skip_provider_button = true
oidc_groups_claim = "my:zitadel:grants"

upstreams = [
    "static://202"
]
email_domains = [
    "*"
]
whitelist_domains = [
    ".<domain>",
    "<domain>",
]
cookie_domains = [
    ".<domain>",
    "<domain>"
]
cookie_csrf_per_request = true
cookie_csrf_expire = "15m"
cookie_secure =  true
cookie_httponly = true
cookie_samesite = "lax"
cookie_secret = "<cookied_secret>"
```

Reply:
There's a couple of issues with `/oauth2/auth` in traefik

1.  When traefik gets the unauthorized response back there is no good way to handle the signin redirect from the forewardAuth middleware so that configuration requires the use of the errors middleware, 
2. The errors middleware does not change its response to match the upstream so instead of a 302 redirect you get a 401 when oauth2-proxy wants to redirect, 
3. Traefik also does not provide the redirect header to oauth2-proxy unless you turn on insecure headers which would allow malicious users to provide arbitrary headers. 
4. The final issue it has is {url} only contains the path and does not contain the full domain so it can't redirect back to the original page authentication happened from 

Now an alternative to this would be to add a new endpoint that properly supports forward auth how traefik expects it with a redirect to sign on instead of just an unauthorized status but the static upstream configuration from the documentation already allows us to implement that in the existing functionality. It's just missing the constraints the `/oauth2/auth` endpoint has, so this pr extends the existing `/` endpoint to allow those constraints to apply.

Reply:
I get the idea behind the PR. I just don't see how your addition of the infinite redirect handling adds the necessary functionality to the root path. 🤔

I'll have to debug the code flow to understand what's actually going on

Reply:
Thanks for providing your setup config. Will make my life a lot easier

Reply:
>  I just don't see how your addition of the infinite redirect handling

That comment came from the existing Auth endpoint config. It's justifying why it's an immediate return and not the auth redirect. On closer looking, I don't think that comment is relevant here I'm going to rewrite it.

Reply:
To head off one question I initially had here's some more notes:

At first glance this check almost appears to belong inside of `OAuthProxy.getAuthenticatedSession` to make it apply to all endpoints that call it, but the `sign_out` and `userinfo` endpoints call that method as well which is why I didn't place it there and instead opted to add the check to just the Proxy handler as it would break those endpoints in potentially frustrating to debug ways.